File: cli/packages/sdk/src/bucket.ts
// TS port of `dif_core::bucket`. Must produce byte-identical buckets for
// every entry in `crates/dif-core/tests/fixtures/bucket_tests.json` — the
// fixture test in `bucket.test.ts` enforces that contract.
//
// Algorithm:
//   1. SHA-256 of (salt || user_id)         — salt is 16 raw bytes, user_id UTF-8
//   2. Take the first 2 bytes as a big-endian u16
//   3. Modulo 10_000 → bucket in [0, 10_000)
//
// Variant selection walks variants in declared order, accumulating
// `weight * 100`; first variant whose cumulative crosses the bucket wins.

import { sha256, bytesToHex, hexToBytes } from "./sha256.js";

/** Bucket namespace — must match `dif_core::BUCKET_NAMESPACE`. */
export const BUCKET_NAMESPACE = "dif.sh/v1";

/** UTF-8 encoder. Reused so we don't allocate one per call. */
const TE = new TextEncoder();

/**
 * Compute the deterministic 16-byte salt for an experiment id, returned as
 * a 32-character lowercase hex string. Mirrors `salt_for` in Rust.
 *
 * The runtime SDK does **not** call this — the generated `client.ts`
 * embeds the precomputed salt. This export exists for the fixture test
 * and for tooling that wants to verify the embed.
 */
export function saltFor(experimentId: string): string {
  const ns = TE.encode(BUCKET_NAMESPACE);
  const id = TE.encode(experimentId);
  const buf = new Uint8Array(ns.length + id.length);
  buf.set(ns);
  buf.set(id, ns.length);
  const digest = sha256(buf);
  return bytesToHex(digest.subarray(0, 16));
}

/**
 * Compute the bucket (0..9999) for a user under the given salt.
 *
 * @param saltHex 32-character lowercase hex. Embedded in the generated file.
 * @param userId  user id string; UTF-8 encoded before hashing.
 */
export function bucket(saltHex: string, userId: string): number {
  const salt = hexToBytes(saltHex);
  if (salt.length !== 16) {
    throw new Error(`expected 16-byte salt, got ${salt.length}`);
  }
  const user = TE.encode(userId);
  const buf = new Uint8Array(salt.length + user.length);
  buf.set(salt);
  buf.set(user, salt.length);
  const digest = sha256(buf);
  const pair = ((digest[0]! << 8) | digest[1]!) >>> 0;
  return pair % 10_000;
}

/**
 * Pick the variant corresponding to a bucket given the experiment's declared
 * variants and weights. Returns `null` if weights don't sum to 100 (which
 * `dif validate` should have caught at build time).
 */
export function selectVariant(
  variants: readonly string[],
  weights: Record<string, number>,
  bucketValue: number,
): string | null {
  let cumulative = 0;
  for (const variant of variants) {
    const w = weights[variant] ?? 0;
    cumulative += w * 100;
    if (bucketValue < cumulative) return variant;
  }
  return null;
}

File: cli/crates/dif-core/src/bucket.rs
//! Deterministic bucketing — Rust side of the cross-language fixture.
//!
//! The TS SDK reimplements this exactly. Any drift breaks the contract; CI
//! runs both against `bucket_tests.json` on every PR.
//!
//! Algorithm:
//!   1. SHA-256 of (salt || user_id)
//!   2. Take the first 4 bytes as a big-endian u32
//!   3. Modulo 10_000 → bucket in [0, 10_000)
//!
//! Four bytes, not two: 2^16 % 10_000 = 5_536, so a u16 source would give
//! buckets 0..=5535 seven preimages and the rest six — a 53.4/46.6 split on a
//! nominal 50/50 experiment. With a u32 source the residual bias is ~1.7e-6.
//!
//! Variant selection walks variants in declared order, accumulating
//! `weight * 100`; the first variant whose cumulative crosses the bucket wins.

use sha2::{Digest, Sha256};

use crate::spec::Variant;
use crate::BUCKET_NAMESPACE;

/// Compute the deterministic bucket salt for an experiment id.
///
/// `salt = SHA256("dif.sh/v1" || experiment_id)[..16]`
pub fn salt_for(experiment_id: &str) -> [u8; 16] {
    let mut hasher = Sha256::new();
    hasher.update(BUCKET_NAMESPACE.as_bytes());
    hasher.update(experiment_id.as_bytes());
    let digest = hasher.finalize();
    let mut salt = [0u8; 16];
    salt.copy_from_slice(&digest[..16]);
    salt
}

/// Compute the bucket (0..10_000) for a user under the given salt.
pub fn bucket(salt: &[u8; 16], user_id: &str) -> u16 {
    let mut hasher = Sha256::new();
    hasher.update(salt);
    hasher.update(user_id.as_bytes());
    let digest = hasher.finalize();
    let quad = u32::from_be_bytes([digest[0], digest[1], digest[2], digest[3]]);
    (quad % 10_000) as u16
}

/// Pick the variant id corresponding to a bucket given the experiment's
/// declared variants. Returns `None` if weights do not sum to 100 (which
/// `validate` should have already caught).
pub fn select_variant(variants: &[Variant], bucket: u16) -> Option<&str> {
    let mut cumulative: u32 = 0;
    for v in variants {
        cumulative += u32::from(v.weight) * 100;
        if u32::from(bucket) < cumulative {
            return Some(v.id.as_str());
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn salt_is_deterministic() {
        assert_eq!(salt_for("checkout-cta-v2"), salt_for("checkout-cta-v2"));
    }

    #[test]
    fn salt_differs_by_experiment() {
        assert_ne!(salt_for("a"), salt_for("b"));
    }

    #[test]
    fn bucket_in_range() {
        let salt = salt_for("checkout-cta-v2");
        for i in 0..100 {
            let b = bucket(&salt, &format!("u_{i}"));
            assert!(b < 10_000);
        }
    }

    /// Pins the modulo-bias fix: with a u16 source a 50/50 split allocated
    /// ~53.4/46.6. 100k synthetic users must land within ±1% of even.
    #[test]
    fn bucket_distribution_is_unbiased() {
        let salt = salt_for("distribution-check");
        let n = 100_000u32;
        let below_split = (0..n)
            .filter(|i| bucket(&salt, &format!("user_{i}")) < 5_000)
            .count();
        let share = below_split as f64 / n as f64;
        assert!(
            (share - 0.5).abs() < 0.01,
            "50/50 split allocated {share:.4} to the first half"
        );
    }
}
